packages/flutter/lib/src/painting/image_provider.dart

```diff
@@ -280,13 +280,23 @@ abstract class ImageProvider<T> {
         }
       );
     }
-    obtainKey(configuration).then<void>((T key) {
-      obtainedKey = key;
-      final ImageStreamCompleter completer = PaintingBinding.instance.imageCache.putIfAbsent(key, () => load(key), onError: handleError);
-      if (completer != null) {
-        stream.setCompleter(completer);
-      }
-    }).catchError(handleError);
+
+    /// `obtainKey` can throw both sync and async errors.
+    /// `catchError` handles cases where async errors are thrown and the outer try block is for sync errors.
+    ///
+    /// `onError` callback on [ImageCache] handles the cases where `obtainKey` is a sync future and `load` throws.
+    try {
```

Comment:
Instead of wrapping everything in a try-catch, just wrap the invocation of obtainKey.

```
Future<T> obtainedKey;
try {
  obtainedKey = obtainKey(configuration);
}
```

Reply:
There was a variable `obtainedKey` already in the scope, so i used `key` instead.